File: Assets/Plugins/iOS/PlayPhoneMultiNet/unity/providers/MNGameSettingsProviderUnity.m
//
//  MNGameSettingsProviderUnity.m
//  PlayPhone SDK for Unity
//
//  Copyright (c) 2012 PlayPhone Inc. All rights reserved.
//

#import "MNDirect.h"
#import "MNGameSettingsProvider.h"

#import "MNGameSettingsProviderUnity.h"

@interface MNGameSettingsProviderUnity : NSObject <MNGameSettingsProviderDelegate>
+ (id)shared;
- (void)onGameSettingListUpdated;
@end

static MNGameSettingsProviderUnity *MNGameSettingsProviderUnityInstance = nil;

@implementation MNGameSettingsProviderUnity

+(id)shared {
    if (MNGameSettingsProviderUnityInstance == nil) {
        MNGameSettingsProviderUnityInstance = [[MNGameSettingsProviderUnity alloc]init];
    }
    
    return MNGameSettingsProviderUnityInstance;
}

- (void)onGameSettingListUpdated {
    [MNUnity callUnityFunction:@"MNUM_onGameSettingListUpdated" withParams:nil];
}
@end

EXTERN_C
const char* _MNGameSettingsProvider_GetGameSettingList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect gameSettingsProvider] getGameSettingList]] UTF8String]);
}

EXTERN_C
const char* _MNGameSettingsProvider_FindGameSettingById(int gameSetId) {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect gameSettingsProvider] findGameSettingById:gameSetId]] UTF8String]);
}

EXTERN_C
bool _MNGameSettingsProvider_IsGameSettingListNeedUpdate() {
    MARK;

    return [[MNDirect gameSettingsProvider] isGameSettingListNeedUpdate];
}

EXTERN_C
void _MNGameSettingsProvider_DoGameSettingListUpdate() {
    MARK;

    [[MNDirect gameSettingsProvider] doGameSettingListUpdate];
}

EXTERN_C
bool _MNGameSettingsProvider_RegisterEventHandler(){
    MARK;

    if ([MNDirect gameSettingsProvider] == nil) {
        return false;
    }

    [[MNDirect gameSettingsProvider] addDelegate:[MNGameSettingsProviderUnity shared]];

    return true;
}

EXTERN_C
bool _MNGameSettingsProvider_UnregisterEventHandler(){
    MARK;

    if ([MNDirect gameSettingsProvider] == nil) {
        return false;
    }

    [[MNDirect gameSettingsProvider] removeDelegate:[MNGameSettingsProviderUnity shared]];

    return true;
}

File: Assets/Plugins/Android/src/com/playphone/multinet/unity/MNGameSettingsProviderUnity.java
package com.playphone.multinet.unity;

import com.playphone.multinet.*;
import com.unity3d.player.UnityPlayer;

public class MNGameSettingsProviderUnity {
    public static void shutdown() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getGameSettingsProvider().shutdown();
            }
        });
    }

    public static String getGameSettingList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getGameSettingsProvider().getGameSettingList());
    }

    public static String findGameSettingById(final int gameSetId) {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getGameSettingsProvider().findGameSettingById(gameSetId));
    }

    public static boolean isGameSettingListNeedUpdate() {
        MNUnity.MARK();

        return MNDirect.getGameSettingsProvider().isGameSettingListNeedUpdate();
    }

    public static void doGameSettingListUpdate() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getGameSettingsProvider().doGameSettingListUpdate();
            }
        });
    }

    protected static class MNGameSettingsProviderEventHandler implements com.playphone.multinet.providers.MNGameSettingsProvider.IEventHandler {

        @Override
        public void onGameSettingListUpdated() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onGameSettingListUpdated");
                }
            });
        }
    }

    private static MNGameSettingsProviderEventHandler registeredEventHandler = null;
    public static synchronized boolean registerEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getGameSettingsProvider() == null) {
            MNUnity.ELog("MNGameSettingsProvider is not ready. Use MNDirect\'s sessionReady event for adding your delegates.");
            return false;
        }

        if (registeredEventHandler == null) {
            registeredEventHandler = new MNGameSettingsProviderUnity.MNGameSettingsProviderEventHandler();
            MNDirect.getGameSettingsProvider().addEventHandler(registeredEventHandler);
        }

        return true;
    }
    
    public static synchronized boolean unregisterEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getGameSettingsProvider() == null) {
            return false;
        }

        if (registeredEventHandler != null) {
            MNDirect.getGameSettingsProvider().removeEventHandler(registeredEventHandler);
            registeredEventHandler = null;
        }

        return true;
    }
}

